runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java

```diff
@@ -172,15 +175,27 @@ public void commit() {
      * Otherwise this throws a TransactionAbortedException.
      *
      * @param timestamp Timestamp to commit the transaction on.
+     * @param numRetries number of times to retry the transaction on internal errors.
      */
-    public void commit(final Timestamp timestamp) {
-        this.timestamp = timestamp;
-        try {
-            txBegin();
-            operations.forEach(Runnable::run);
-        } finally {
-            txEnd();
-            operations.clear();
+    public void commit(final Timestamp timestamp, int numRetries) {
+        if (numRetries <= 0) {
+            throw new IllegalArgumentException("TxBuilder::commit: numRetries <=0"+numRetries);
+        }
+        while (numRetries-- > 0) {
```

Comment:
And also please don't reduce the variable and check the result at the same time. It's confusing and makes code more error prone

Reply:
done. will use separate variable.